### PR TITLE
Invite sending and tracking

### PR DESCRIPTION
**Pull request by timabell**
### Context

We want to send invites to users and track when they were sent

### Changes proposed in this pull request

* Add a new unit-tested service to orchestrate the behaviour.
* Add new storage for recording date of invite.
* Wire up "invite" API end point (requires pre-shared api key)

### CODE REVIEW

File: src/ManageCourses.Api/Services/InviteService.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        public void Invite(string email)
+        {
+            var mcUser = _context.McUsers.ByEmail(email).SingleOrDefault();
+            if (mcUser == null)
+            {
+                throw new McUserNotFoundException();
+            }
+            var inviteEmailModel = new InviteEmailModel(mcUser);
+            _inviteEmailService.Send(inviteEmailModel);
+            if (mcUser.InviteDateUtc == null)
```
Comment (defong) [inline src/ManageCourses.Api/Services/InviteService.cs]: HasValue?

File: tests/ManageCourses.Tests/SmokeTests/SystemUnderTest.cs [hunk 2]
```diff
@@ -106,7 +106,7 @@ public void DataExport_badAccesCode_404()
             var client = new ManageCoursesApiClient(new MockApiClientConfiguration(accessToken), httpClient);
             client.BaseUrl = localWebHost.Address;
 
-            var result = await client.Invite_IndexAsync();
+            var result = await client.Invite_IndexAsync("wibble");
```
Comment (defong) [inline tests/ManageCourses.Tests/SmokeTests/SystemUnderTest.cs]: I thought it was an email address?

File: src/ManageCourses.Api/Controllers/InviteController.cs [hunk 3]
```diff
@@ -1,25 +1,26 @@
...
-        public InviteController(IInviteEmailService inviteEmailService)
+        public InviteController(IInviteService inviteService)
         {
-            _inviteEmailService = inviteEmailService;
+            _inviteService = inviteService;
         }
 
         [HttpPost]
-        public IActionResult Index() 
+        public IActionResult Index(string email)
```
Comment (defong) [inline src/ManageCourses.Api/Controllers/InviteController.cs]: https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-2.1

[EmailAddress]

File: tests/ManageCourses.Tests/SmokeTests/SystemUnderTest.cs [hunk 4]
```diff
@@ -106,7 +91,10 @@ public void DataExport_badAccesCode_404()
             var client = new ManageCoursesApiClient(new MockApiClientConfiguration(accessToken), httpClient);
             client.BaseUrl = localWebHost.Address;
 
-            var result = await client.Invite_IndexAsync();
+            const string email = "feddie.krueger@example.org";
+            client.Data_ImportReferenceDataAsync(TestData.MakeReferenceDataPayload(email)).Wait();
```
Comment (defong) [inline tests/ManageCourses.Tests/SmokeTests/SystemUnderTest.cs]: whats `.Wait()`?

Can it not be `.Result` ?

Comment (defong) [reply tests/ManageCourses.Tests/SmokeTests/SystemUnderTest.cs]: What happened to your test method as a whole?

**Comment by timabell** [comment]
Test failures need sorting out. Will re-open in some form later.

**Pull request closed without merging**